test/src/lib/backoff.spec.ts

```diff
@@ -0,0 +1,163 @@
+import { expect } from 'chai';
+
+import Failer from '../../lib/failer';
+
+import { StatusError } from '../../../src/lib/errors';
+import { withBackoff, OnRetryOpts } from '../../../src/lib/backoff';
+
+describe('lib/backoff', async () => {
+	const DEFAULT_OPTIONS = {
+		maxRetries: 5,
+		maxDelay: 50,
+		minDelay: 1,
+	};
+
+	it('resolves after 3 retries', async () => {
+		// Create a function that will fail 3 times so it succeeds on the 4th
+		const failer = new Failer(3);
+		// Wrap function withBackoff
+		const fnWithBackoff = withBackoff(async () => {
+			return failer.willResolve('data that fails a lot...');
+		}, DEFAULT_OPTIONS);
```

Comment:
I have to call the failer inside a closure because for whatever reason if I don't the references to this inside the class are undefined. I spent some time debugging it and couldn't figure it out.

Reply:
@20k-ultra Could you clarify: the references to what are undefined?

Reply:
The failer class has functions that increment an internal counter for number of fails. When this `fnWithBackoff` runs and tries to do `this.count++` the reference to this is undefined. I hope that makes sense.

Reply:
As a theory, it is possible that when the Promise is loaded into the event loop, the object context is no longer available. Maybe doing something like the following might work? Using a closure sounds fine imho too

```
	public async willResolve(resolvesWith: string): Promise<string> {
                const that = this; 
		return new Promise((resolve, reject) => {
			if (++that.counter <= that.maxFails) {
				return reject('Not ready!');
			}
			return resolve(resolvesWith);
		});
	}
```

Reply:
I have seen that = this pattern used before for situations like this but that still doesn't work :/

Reply:
For future reference, when referring to the JS/TS keyword `this` could we wrap that in backticks for clarity? It's too ambiguous of a word otherwise

Reply:
@20k-ultra What is the non-wrapped test code that you tried before, that doesn't work?